docs/architecture/adr-111-nil-mempool.md

```diff
@@ -0,0 +1,239 @@
+* `create_empty_blocks`, in `consensus` section, is set to `false`.
+
+### Impacted Workflows
+
+* *Submitting a transaction*. Users are not to submit transactions via CometBFT's RPC.
+  `BroadcastTx*` RPC methods will fail.
+  The application running on a full node must offer an interface for users to submit new transactions.
+  It could also be a distinct node (or set of nodes) in the network.
+  These considerations are exclusively the application's concern in this approach.
+* *Time to propose a block*. The consensus reactor will call `ReapMaxBytesMaxGas` which will return a `nil` slice.
+  `RequestPrepareProposal` will thus contain no transactions.
+* *Consensus waiting for transactions to become available*. `TxsAvailable()` returns a closed channel,
+  so consensus doesn't block (potentially producing empty blocks).
+  At any rate, a configuration with the `nil` mempool and `create_empty_blocks` set to `false`
+  will be rejected at the first place.
+* *A new block is decided*.
+  * When `Update` is called, nothing is done (no decided transaction is removed).
+  * Locking and unlocking the mempool has no effect.
+
+### Impact on Current Release Plans
+
+The changes needed for this approach, are fairly simple, and the logic is clear.
+This might allow us to even deliver it as part of CometBFT `v1` (our next release)
+even without a noticeable impact on `v1`'s delivery schedule.
+
+The CometBFT team (learning from past dramatic events) usually takes a conservative approach
+for backporting changes to release branches that have already undergone a full QA cycle
+(and thus are in code-freeze mode).
+For this reason, although the limited impact of these changes would limit the risks
+of backporting to `v0.38.x` and `v0.37.x`, a careful risk/benefit evaluation will
+have to be carried out.
+
+Backporting to `v0.34.x` does not make sense as this version predates the release of `ABCI 1.0`,
+so using the `nil` mempool renders CometBFT's operation useless.
+
+>TODO: Need to add a small section here explaining the dilemma between a Consensus Param and a
+>      `config.toml` field.
+
+## Consequences
+
+### Positive
+
+- Applications can now find mempool mechanisms that fit better their particular needs:
+  - Ad-hoc ways to add, remove, merge, reorder, modify, prioritize transactions according
+    to application needs.
+  - A way to disseminate transactions (gossip-based or other) to get the submitted transactions
+    to proposers. The application developers can devise simpler, efficient mechanisms tailored
+    to their application.
+  - Back-pressure mechanisms to prevent malicious users from abusing the transaction
+    dissemination mechanism.
+- In this approach, CometBFT's peer-to-peer layer does not need to deal with transaction gossipping,
+  and its resources can be used by other reactors such as consensus, evidence,
+  block-sync, or state-sync.
+- There is no risk for the operators of a network to provide inconsistent configurations
```

Comment:
One scenario not clear to me is that if the application implements its ad-hoc mechanism, and some operators configured the `nil` mempool but others might not configure it (e.g. configure the CometBFT regular one). Will that cause any issues ? I think it should be OK because the app should handle the transaction dissemination, but just wondering if there are possible scenarios where a mis-configuration would cause side-effects or issues.

Reply:
Good point 👍 We must ensure the node doesn't panic when receiving a slice of transactions via the p2p network.

`RequestPrepareProposal` will contain a list of transactions in such case, but as long as the app ignores it and uses a custom mempool, it should be okay.

Reply:
The way I see it, in this particular case, there will be a risk of wasting resources by those nodes that have a misconfigured mempool (bandwidth, CPU, memory, etc) if they are submitted TXs via CometBFT's RPC.
It could get nasty, if there are TXs submitted (incorrectly) via CometBFT's RPC, but those TXs are never submitted (correctly) to the App. As those TXs risk being there for a long time.